### PR TITLE
Varbits

### PR DESCRIPTION
**Pull request by Zac343**

### CODE REVIEW

**Comment by kyle-escobar** [comment]
Things to change:

- BitManipulation class belong in the Util module.
- VarpSet belongs in the engine. An api interface will need to be created and override methods which are need for api interaction.
- VarpSet is currently exactly rsmod's code. You also need to change the package to the correct one in that class.
- Remove OSRSGameFrame. Varps and varbits are defined in the api module in const/varbits.kt.
- Varps.kt is not needed since varbits are a bitwise shift of 3 from the startbit to the end bit. The start and end bits can be pulled from the cache for a given varbit.

All the game logic for varps should be in the engine. A interface class gets implemented into that and overrides functions / values that are needed for api use.
That way we can separate the api from the game logic for plugin development.

Packets look good.
Saving of run energy looks good.

File: api/src/main/kotlin/io/rsbox/api/consts/Varbit.kt [hunk 1]
```diff
@@ -3,7 +3,7 @@ package io.rsbox.api.consts
 object Varbit {
     const val XP_DROPS_VISIBLE = 4702
     const val DISABLE_MOUSEWHEEL_ZOOM = 6357
-    const val HIDE_PM_WHEN_CHAT_HIDDEN = 4089
+    const val HIDE_PM_WHEN_CHAT_HIDD4702EN = 4089
```
Comment (Desetude) [inline api/src/main/kotlin/io/rsbox/api/consts/Varbit.kt]: I guess this was accidental?

File: api/build.gradle [hunk 2]
```diff
@@ -9,11 +9,12 @@ shadowJar {
     baseName = "rsbox-api"
     //noinspection GrDeprecatedAPIUsage
     classifier = null
-    
+
     dependencies {
+        implementation project(':util')
```
Comment (kyle-escobar) [inline api/build.gradle]: Why was this needed??

File: engine/src/main/kotlin/io/rsbox/engine/model/entity/Player.kt [hunk 3]
```diff
@@ -139,6 +143,55 @@ class Player(val channel: Channel) : LivingEntity(), io.rsbox.api.entity.Player
...
+        val def = _world.engine._cache.get(CacheVarbit::class.java, id)
+        varps.setBit(def.varp, def.startBit, def.endBit, value)
+        sendvarbit()
+    }
+
+    
+    /*
+     * Not sure if we want it on cycle like rs mod or every time you send the varbit so just added this for now
+     */
+    fun sendvarbit() {
```
Comment (kyle-escobar) [inline engine/src/main/kotlin/io/rsbox/engine/model/entity/Player.kt]:
```suggestion
    fun sendVarbits() {
```

File: engine/src/main/kotlin/io/rsbox/engine/net/packet/impl/server/RunClientScriptPacket.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+class RunClientScriptPacket(val id: Int, vararg val args: Any) : ServerPacket() {
+    override var opcode = 62
+    override var type = PacketType.VARIABLE_SHORT
+
+    override fun encode() {
+        initPacket()
+
+        packet.put(type = DataType.INT, value = id)
+
```
Comment (kyle-escobar) [inline engine/src/main/kotlin/io/rsbox/engine/net/packet/impl/server/RunClientScriptPacket.kt]: Packet structure for RunClientScript is encoded on the client as follows below:
```
			if (ServerPacket.field2177 == buf.serverPacket) {
				var37 = packet.readStringCp1252NullTerminated();
				Object[] var57 = new Object[var37.length() + 1];

				for (var17 = var37.length() - 1; var17 >= 0; --var17) {
					if (var37.charAt(var17) == 's') {
						var57[var17 + 1] = packet.readStringCp1252NullTerminated();
					} else {
						var57[var17 + 1] = new Integer(packet.readInt());
					}
				}

				var57[0] = new Integer(packet.readInt());
				ScriptEvent var48 = new ScriptEvent();
				var48.args = var57;
				LoginPacket.runScriptEvent(var48);
				buf.serverPacket = null;
				return true;
			}
```

So encoding would be something like this.
```
override fun encode() {
    initPacket()

    // Types
    packet.putBytes(encodeTypes())
    packet.putBytes(encodeArgs())
    packet.put(type = DataType.INT, order = DataOrder.BIG, transformation = DataTransformation.NONE, value = scriptId)
}

private fun encodeTypes(): ByteArray {
    val types = CharArray(message.args.size + 1)
    for(i in 0 until args.size) {
        types[i] = when(args[i]) {
            is String -> 's'
            else -> 'i'
        }
        types[args.size] = 0.toChar()
        return String(types).toByteArray()
}

private fun encodeArgs(): ByteArray {
    val args = mutableListOf<Byte>()

    for(i in args.size - 1 downTo 0) {
        val v = args[i]
        when(v) {
            is String -> {
                v.toByteArray().forEach { args.add(it) }
                args.add(0)
            }

           is Number -> {
               args.add((v.toInt() shr 24).toByte())
               args.add((v.toInt() shr 16).toByte())
               args.add((v.toInt() shr 8).toByte())
              args.add(v.toByte())
        }
}
```

This is just psudo code and probably isnt right.

File: game/build.gradle [hunk 5]
```diff
@@ -2,4 +2,5 @@ plugins {}
 
 dependencies {
     api project(':api')
-}
\ No newline at end of file
+    implementation project(':util')
```
Comment (kyle-escobar) [inline game/build.gradle]: Why was this added??? We in theory should not need Util classes in the Game module as the game module interfaces with the API module. All logic (which the Util module holds helper classes for) is dont in the engine module.

File: game/src/main/kotlin/io/rsbox/game/Game.kt [hunk 6]
```diff
@@ -13,8 +14,26 @@ object Game {
...
 
         InterfaceDestination.values.filter { pane -> pane.interfaceId != -1 }.forEach { pane ->
+            if (pane == InterfaceDestination.XP_COUNTER && player.getVarbit(Varbit.XP_DROPS_VISIBLE) == 0) {
+                return@forEach
+            } else if (pane == InterfaceDestination.MINI_MAP && player.getVarbit(Varbit.HIDE_DATA_ORBS) == 1) {
+                return@forEach
+            }
             player.openInterface(pane.interfaceId, pane)
         }
+
```
Comment (kyle-escobar) [inline game/src/main/kotlin/io/rsbox/game/Game.kt]: Please remove the displayName setting. I have plans for on initial login, the set display name interface from OSRS will appear.

**Comment by kyle-escobar** [comment]
I am closing this PR since they way I plan to handle varps and varbit is different to how RSMod handles them currently. I want to have them automatically build based on the start / end bit read from the game cache. These will be tied into player attributes.

You are welcome to make a PR with the packets you have added however varbits are going to be handled differently and through a more modular system.
